src/power_grid_model/core/power_grid_model.py

```diff
@@ -149,92 +157,95 @@ def get_indexer(self, component_type: str, ids: np.ndarray):
         assert_no_error()
         return indexer
 
-    # pylint: disable=too-many-locals
-    # pylint: disable=too-many-branches
-    # pylint: disable=too-many-arguments
-    def _calculate(
+    def _output_type(self, options: Options):
+        return get_output_type(calculation_type=options.calculation_type, symmetric=options.symmetric)
+
+    def _construct_output(
         self,
+        output_component_types: Optional[Union[Set[str], List[str]]],
+        options: Options,
+        batch_size: int,
+    ) -> Dict[str, np.ndarray]:
+        # prepare result dataset
+        all_component_count = self.all_component_count
+
+        # for power flow, there is no need for sensor output
+        if options.calculation_type == CalculationType.power_flow:
+            all_component_count = {k: v for k, v in all_component_count.items() if "sensor" not in k}
+
+        # limit all component count to user specified component types in output
+        if output_component_types is None:
+            output_component_types = set(all_component_count.keys())
+
+        return create_output_data(
+            output_component_types=output_component_types,
+            output_type=self._output_type(options=options),
+            all_component_count=all_component_count,
+            batch_size=batch_size,
+        )
+
+    # pylint: disable=too-many-arguments
+    @staticmethod
+    def _options(
         calculation_type: CalculationType,
         symmetric: bool,
         error_tolerance: float,
         max_iterations: int,
         calculation_method: Union[CalculationMethod, str],
-        update_data: Optional[Dict[str, Union[np.ndarray, Dict[str, np.ndarray]]]],
         threading: int,
+    ) -> Options:
+        if isinstance(calculation_method, str):
+            calculation_method = CalculationMethod[calculation_method]
+
+        opt = Options()
+        opt.calculation_type = calculation_type.value
```

Comment:
here you just do 
```python
for k, v in kw_options.items():
    setattr(opt, k, v)
```

Reply:
## option 1

```py
class Options:
    """
    Option class
    """

    _opt: OptionsPtr
    # option setter
    calculation_type = OptionSetter(pgc.set_calculation_type)
    calculation_method = OptionSetter(pgc.set_calculation_method)
    symmetric = OptionSetter(pgc.set_symmetric)
    error_tolerance = OptionSetter(pgc.set_err_tol)
    max_iterations = OptionSetter(pgc.set_max_iter)
    threading = OptionSetter(pgc.set_threading)
```

## option 2a

```py
    @staticmethod
    def _options(**kwargs) -> Options:
        calculation_method = kwargs.get("calculation_method")
        if calculation_method is not None and isinstance(calculation_method, str):
            kwargs["calculation_method"] = CalculationMethod[calculation_method]

        if "max_iterations" in kwargs:
            kwargs["max_iteration"] = kwargs["max_iterations"]

        opt = Options()
        for key, value in kwargs.items():
            setattr(opt, key, value.value if isinstance(value, IntEnum) else value)
        return opt
```


## option 2b

```py
class Options:
    """
    Option class
    """

    _opt: OptionsPtr
    # option setter
    calculation_type = OptionSetter(pgc.set_calculation_type)
    calculation_method = OptionSetter(pgc.set_calculation_method)
    symmetric = OptionSetter(pgc.set_symmetric)
    error_tolerance = OptionSetter(pgc.set_err_tol)
    max_iterations = OptionSetter(pgc.set_max_iter)
    threading = OptionSetter(pgc.set_threading)
    max_iteration = max_iterations
```

Reply:
No, that's nothing to do with C-API. It's just the name of the descriptor. You can change it. Everything inside `core` is not public Python API, unless imported at root level.

Reply:
no the name in the C-API is completely different:

```c
/**
 * @brief Specify maximum number of iterations. Only applicable if using iterative method.
 *
 * @param handle
 * @param opt pointer to option instance
 * @param max_iter Maximum number of iterations
 */
PGM_API void PGM_set_max_iter(PGM_Handle* handle, PGM_Options* opt, PGM_Idx max_iter);
```

```cpp
// options
struct PGM_Options {
    Idx calculation_type{PGM_power_flow};
    Idx calculation_method{PGM_default_method};
    Idx symmetric{1};
    double err_tol{1e-8};
    Idx max_iter{20};
    Idx threading{-1};
};
```

Reply:
> Go for Option 1

ok 😄 that has my preference as well